### PR TITLE
Update CMake scripts to include the ADIOS2 IO backend

### PR DESCRIPTION
**Pull request by mathrack**

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -74,10 +74,16 @@ if (ENABLE_PROFILER)
     target_link_libraries(decomp2d PRIVATE caliper)
   endif()
 endif()
+
 if (HALO_DEBUG)
   target_compile_definitions(decomp2d PUBLIC -DHALO_DEBUG)
 endif()
 
+if (IO_BACKEND MATCHES "adios2")
+  add_definitions("-DADIOS2")
```
Comment (mathrack) [inline src/CMakeLists.txt:83]: Should be `target_compile_definitions(decomp2d PUBLIC -DADIOS2)` ?

File: cmake/D2D_Compilers.cmake [hunk 2]
```diff
@@ -97,3 +97,16 @@ option(SINGLE_PRECISION_OUTPUT "Build XCompact with output in single precision"
 if (SINGLE_PRECISION_OUTPUT)
   add_definitions("-DSAVE_SINGLE")
```
Comment (mathrack) [inline cmake/D2D_Compilers.cmake:98]: This should be moved and changed to `target_compile_definitions(decomp2d PUBLIC -DSAVE_SINGLE)` ?

Comment (mathrack) [reply cmake/D2D_Compilers.cmake:98]: Some preprocessor variables are used in the examples. We can move some preprocessor variables to `src/CMakeLists.txt` but not all of them. If we move all of them, we should add the variables in the CMakeLists.txt of the example

Comment (rfj82982) [reply cmake/D2D_Compilers.cmake:98]: Got the following error when compiling 
```
Building Fortran object src/CMakeFiles/decomp2d.dir/io.f90.o
cd /Users/rfj82982/GIT_projects/2decomp-fft_cedric/build/src && /opt/local/bin/mpif90-mpich-gcc12 -DADIOS2 -DADIOS2_USE_MPI -DDOUBLE_PREC -DVERSION=\"e85638f\" -I/Users/rfj82982/GIT_projects/2decomp-fft_cedric/src -I/Users/rfj82982/GIT_projects/2decomp-fft_cedric/build/include -I/opt/local/include/mpich-mp -I/Users/rfj82982/GIT_projects/ADIOS2/build/opt/include/adios2/fortran -cpp -std=f2008 -ffree-line-length-none -fallow-argument-mismatch -O3 -march=native -J../include -c /Users/rfj82982/GIT_projects/2decomp-fft_cedric/src/io.f90 -o CMakeFiles/decomp2d.dir/io.f90.o
/Users/rfj82982/GIT_projects/2decomp-fft_cedric/src/io.f90:140:91:

  140 |       call adios2_init(adios, trim(config_file), decomp_2d_comm, adios2_debug_mode, ierror)
      |                                                                                           1
Error: There is no specific subroutine for the generic 'adios2_init' at (1)
make[2]: *** [src/CMakeFiles/decomp2d.dir/io.f90.o] Error 1
make[1]: *** [src/CMakeFiles/decomp2d.dir/all] Error 2
make: *** [all] Error 2
```
I’ll check if i can fix it.

Comment (rfj82982) [reply cmake/D2D_Compilers.cmake:98]: I need to check this ADIOS 2.9. I am on this repo 
https://github.com/ornladios/ADIOS2.git
on the main branch. Just did a pull to have the most recent version and build/install. 

I cannot find adios2_init among the stuffs in the 
include/adios2/fortran

Comment (rfj82982) [reply cmake/D2D_Compilers.cmake:98]: I moved from the master to the 2.9.0 branch and I have also tested the tar file. In all cases I always get the error above.

Comment (rfj82982) [reply cmake/D2D_Compilers.cmake:98]: It looks like that the arguments for `adios2_init` are now different from what we have in our call 
https://adios2.readthedocs.io/en/latest/api_full/api_full.html#fortran-bindings
```
! MPI versions
subroutine adios2_init(adios, comm, ierr)
subroutine adios2_init(adios, config_file, comm, ierr)

! Non-MPI serial versions
subroutine adios2_init(adios, ierr)
subroutine adios2_init(adios, config_file, ierr)
```
while we have
`call adios2_init(adios, trim(config_file), decomp_2d_comm, adios2_debug_mode, ierror)`

Comment (mathrack) [reply cmake/D2D_Compilers.cmake:98]: I have tested the present PR against ADIOS version 2.8.

Comment (mathrack) [reply cmake/D2D_Compilers.cmake:98]: https://github.com/ornladios/ADIOS2/pull/3204

**Comment by pbartholomew08** [comment]
This should now build against adios v2.8 and v2.9. Currently it fails some I/O tests (before and after commit 3394db2)

File: cmake/D2D_Compilers.cmake (hunk 2 above)
Comment (mathrack) [reply cmake/D2D_Compilers.cmake:98]: Fixed in https://github.com/2decomp-fft/2decomp-fft/pull/195/commits/3394db2d617b68d7e9a7f9f82a9bbd4460fcb630

**Comment by mathrack** [comment]
I think this PR is complete : the CMake scripts provide the possibility to compile the library with ADIOS2 IO. We just need to select `add_definitions("-DADIOS2")` or `target_compile_definitions(decomp2d PUBLIC -DADIOS2)`.

File: src/CMakeLists.txt (hunk 1 above)
Comment (mathrack) [reply src/CMakeLists.txt]: Fixed in https://github.com/2decomp-fft/2decomp-fft/pull/195/commits/887938756e6dfbb43077fd567c61e73a93bdf24e

**Comment by pbartholomew08** [comment]
I believe this is now ready

File: examples/io_test/io_plane_test.f90 [hunk 3]
```diff
@@ -36,6 +36,9 @@ program io_plane_test
    integer :: xen1, xen2, xen3
    logical :: found
 
+   character(len=*), parameter :: io_name = "test-io"
+   integer, parameter :: output2D = 0
```
Comment (mathrack) [inline examples/io_test/io_plane_test.f90:40]: this is not used ?

**Review by rfj82982 — DISMISSED** [review]

File: cmake/D2D_MPI.cmake [hunk 4]
```diff
@@ -1,6 +1,12 @@
 # MPI CMakeLists
 
-find_package(MPI REQUIRED)
+find_package(MPI REQUIRED COMPONENTS Fortran)
+
+# adios2 IO backend requires C and C++ MPI components
+if (IO_BACKEND MATCHES "adios2")
+    find_package(MPI REQUIRED COMPONENTS C CXX)
```
Comment (pbartholomew08) [inline cmake/D2D_MPI.cmake:7]: Should Fortran not also be a required component?

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -74,10 +74,16 @@ if (ENABLE_PROFILER)
...
   endif()
 endif()
+
 if (HALO_DEBUG)
   target_compile_definitions(decomp2d PUBLIC -DHALO_DEBUG)
 endif()
 
+if (IO_BACKEND MATCHES "adios2")
+  target_compile_definitions(decomp2d PUBLIC -DADIOS2)
+  target_link_libraries(decomp2d PRIVATE adios2::fortran_mpi adios2::fortran)
```
Comment (pbartholomew08) [inline src/CMakeLists.txt:84]: could/should this be `PUBLIC`?

**Review by pbartholomew08 — APPROVED** [review]
Minor comment aside I think this looks fine!

File: src/CMakeLists.txt (hunk 5 above)
Comment (rfj82982) [reply src/CMakeLists.txt:84]: Yes the PUBLIC seems to be the better solution 
No idea why with gnu is OK

File: cmake/D2D_MPI.cmake (hunk 4 above)
Comment (rfj82982) [reply cmake/D2D_MPI.cmake:7]: It is just on the line above. Don’t know if we want to make all 3 required components independently of ADIOS

Comment (pbartholomew08) [reply cmake/D2D_MPI.cmake:7]: Ah, missed that!

**Review by pbartholomew08 — APPROVED** [review]
LGTM!

**Pull request merged**